pkg/daemon/daemon.go

```diff
@@ -245,145 +348,80 @@ func (dn *Daemon) operatorConfigChangeHandler(old, new interface{}) {
-			dn.exitCh <- err
-			return
-		}
-		reqDrain = reqDrain || d
-		reqReboot = reqReboot || r
-	}
-	glog.V(2).Infof("nodeStateAddHandler(): reqDrain %v, reqReboot %v", reqDrain, reqReboot)
-
-	if reqDrain {
-		glog.Info("nodeStateAddHandler(): drain node")
-		dn.drainNode(nodeState.GetName())
-	}
-	for k, p := range dn.LoadedPlugins {
-		if k != GenericPlugin {
-			err := p.Apply()
-			if err != nil {
-				glog.Errorf("nodeStateAddHandler(): plugin %s fail to apply: %v", k, err)
-				dn.exitCh <- err
-				return
-			}
-		}
-	}
-
-	if len(dn.LoadedPlugins) > 1 && !reqReboot {
-		// Apply generic_plugin last
-		err = dn.LoadedPlugins[GenericPlugin].Apply()
-		if err != nil {
-			glog.Errorf("nodeStateAddHandler(): generic_plugin fail to apply: %v", err)
-			dn.exitCh <- err
-			return
-		}
-	}
-
-	if reqReboot {
-		glog.Info("nodeStateAddHandler(): reboot node")
-		rebootNode()
-		return
-	}
-
-	// restart device plugin pod
-	if reqDrain {
-		if err := dn.restartDevicePluginPod(); err != nil {
-			glog.Errorf("nodeStateAddHandler(): fail to restart device plugin pod: %v", err)
-			dn.exitCh <- err
-			return
-		}
-	}
-
-	dn.completeDrain()
-	dn.refreshCh <- Message{
-		syncStatus:    "Succeeded",
-		lastSyncError: "",
-	}
-}
-
-func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
+func (dn *Daemon) nodeStateSyncHandler(generation int64) error {
 	var err, lastErr error
-	newState := new.(*sriovnetworkv1.SriovNetworkNodeState)
-	oldState := old.(*sriovnetworkv1.SriovNetworkNodeState)
-	glog.V(2).Infof("nodeStateChangeHandler(): new generation is %d", newState.GetObjectMeta().GetGeneration())
+	glog.V(0).Infof("nodeStateSyncHandler(): new generation is %d", generation)
 	// Get the latest NodeState
 	var latestState *sriovnetworkv1.SriovNetworkNodeState
 	err = wait.PollImmediate(10*time.Second, 1*time.Minute, func() (bool, error) {
 		latestState, lastErr = dn.client.SriovnetworkV1().SriovNetworkNodeStates(namespace).Get(context.Background(), dn.name, metav1.GetOptions{})
 		if lastErr == nil {
 			return true, nil
 		}
-		glog.Warningf("nodeStateChangeHandler(): Failed to fetch node state %s (%v); retrying...", dn.name, lastErr)
+		glog.Warningf("nodeStateSyncHandler(): Failed to fetch node state %s (%v); retrying...", dn.name, lastErr)
 		return false, nil
 	})
 	if err != nil {
 		if err == wait.ErrWaitTimeout {
-			dn.exitCh <- goerr.Wrapf(lastErr, "nodeStateChangeHandler(): Timed out trying to fetch the latest node state %s", dn.name)
-			return
+			return fmt.Errorf("nodeStateSyncHandler(): Timed out trying to fetch the latest node state: %v", lastErr)
 		}
-		dn.exitCh <- err
-		return
+		return fmt.Errorf("%v: %v", err, lastErr)
 	}
-	latest := latestState.GetObjectMeta().GetGeneration()
-	if latest != newState.GetObjectMeta().GetGeneration() {
-		glog.Errorf("nodeStateChangeHandler(): the lastet generation is %d, skip", latest)
-		return
+	latest := latestState.GetGeneration()
+	if latest != generation {
```

Comment:
This can be moved in the updateHandler, so you end up having only different versions in the queue.
Also, note that the final order you process them might be different from the original one, because if one error happens, the event is put back in the queue and processed after a fresher one, but I don't think this is an issue.

Reply:
If we move it to the updateHandler, we shall do the same thing when putting a failed event back to the queue as well. After all, we don't really care about the key itself, we only need to keep the state synced with the latest. Therefore I prefer to leave this logic here.

Reply:
If you move to `updateHandler` you'll have the guarantee of having only different generations in the queue.
Also, this if `if latest != generation {` may be true if you are triggered by a variation, then you read the current status using the client and you retrieve the very same generation.